File: src/lib/community/rideGroupInvite.ts
/**
 * Teilbarer Einladungslink für „Zusammen raus“.
 * HTTPS: /library?group=<id>&g=<token>
 * App: aetherride://platz?group=<id>&g=<token>
 * Alte Links mit 6-Zeichen-Code bleiben gültig.
 * Code zum Abtippen nur bei öffentlichen / Platz-Gruppen.
 */

import { siteOrigin, appDeepLink } from "@/lib/web/appLinks";
import type { RideGroup } from "@/lib/community/types";
import {
  isRideGroupId,
  isTypedJoinCode,
  normalizeJoinCode,
  RIDE_GROUP_JOIN_CODE_LEN,
} from "@/lib/community/rideGroup";
import { parseTourShareMap } from "@/lib/community/shareCodec";
import type { SharedTourPayload } from "@/lib/community/shareTypes";

export type RideGroupInvitePayload = {
  v: 1;
  kind: "group";
  id: string;
  code: string;
  title: string;
  savedRouteId: string;
  catalogTourId?: string;
  hostUserId?: string;
  start: string;
  end: string;
  tour?: SharedTourPayload;
};

function toBase64Url(json: string): string {
  if (typeof btoa === "function") {
    const b64 = btoa(unescape(encodeURIComponent(json)));
    return b64.replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
  }
  return Buffer.from(json, "utf8")
    .toString("base64")
    .replace(/\+/g, "-")
    .replace(/\//g, "_")
    .replace(/=+$/, "");
}

function fromBase64Url(token: string): string {
  const b64 = token.replace(/-/g, "+").replace(/_/g, "/");
  const pad = b64.length % 4 === 0 ? "" : "=".repeat(4 - (b64.length % 4));
  const full = b64 + pad;
  if (typeof atob === "function") {
    return decodeURIComponent(escape(atob(full)));
  }
  return Buffer.from(full, "base64").toString("utf8");
}

export function encodeGroupInvite(
  group: RideGroup,
  tour?: SharedTourPayload
): string {
  const base: RideGroupInvitePayload = {
    v: 1,
    kind: "group",
    id: group.id,
    code: group.joinCode.trim().toUpperCase(),
    title: group.title,
    savedRouteId: group.savedRouteId,
    catalogTourId: group.catalogTourId,
    hostUserId: group.hostUserId,
    start: group.startWindowStart,
    end: group.startWindowEnd,
  };
  let payload: RideGroupInvitePayload = tour ? { ...base, tour } : base;
  let token = toBase64Url(JSON.stringify(payload));
  if (token.length > 2400 && payload.tour?.track) {
    payload = {
      ...payload,
      tour: { ...payload.tour, includeTrack: false, track: undefined },
    };
    token = toBase64Url(JSON.stringify(payload));
  }
  if (token.length > 2400) {
    token = toBase64Url(JSON.stringify(base));
  }
  return token;
}

export function decodeGroupInvite(
  token: string
): RideGroupInvitePayload | null {
  try {
    const data = JSON.parse(fromBase64Url(token)) as RideGroupInvitePayload;
    if (data?.v !== 1 || data.kind !== "group") return null;
    const code = String(data.code || "")
      .trim()
      .toUpperCase();
    if (code.length !== 6 || !data.id || !data.title || !data.savedRouteId) {
      return null;
    }
    if (!data.start || !data.end) return null;
    const tour = parseTourShareMap(data.tour) ?? undefined;
    return { ...data, code, tour };
  } catch {
    return null;
  }
}

export function groupInviteRef(group: Pick<RideGroup, "id" | "joinCode">): string {
  const id = String(group.id || "").trim();
  if (id) return id;
  return String(group.joinCode || "").trim().toUpperCase();
}

export function groupInvitePath(groupRef: string, token?: string): string {
  const ref = groupRef.trim();
  const q = new URLSearchParams({ group: ref });
  if (token) q.set("g", token);
  return `/library?${q.toString()}`;
}

export function groupInviteHttps(groupRef: string, token?: string): string {
  const path = groupInvitePath(groupRef, token);
  const origin =
    siteOrigin() ||
    (typeof window !== "undefined" ? window.location.origin : "");
  return origin ? `${origin}${path}` : path;
}

export function groupInviteScheme(groupRef: string, token?: string): string {
  const ref = groupRef.trim();
  const q = new URLSearchParams({ group: ref });
  if (token) q.set("g", token);
  return appDeepLink(`platz?${q.toString()}`);
}

export type PastedGroupJoin = { ref: string; token?: string };

/** Paste from WhatsApp / Messages. Prefers the URL that carries `g=`. */
export function parsePastedGroupJoin(raw: string): PastedGroupJoin | null {
  const text = raw.trim();
  if (!text) return null;
  let withoutToken: PastedGroupJoin | null = null;
  const urlRe = /(?:https?:\/\/|aetherride:\/\/)[^\s]+/gi;
  let m: RegExpExecArray | null;
  while ((m = urlRe.exec(text)) !== null) {
    try {
      const uri = new URL(m[0]);
      const group = (uri.searchParams.get("group") || uri.searchParams.get("code") || "").trim();
      if (!group) continue;
      const token = uri.searchParams.get("g")?.trim() || undefined;
      const hit: PastedGroupJoin = { ref: group, token };
      if (hit.token) return hit;
      withoutToken ??= hit;
    } catch {
      /* not a URL */
    }
  }
  if (withoutToken) return withoutToken;
  const compact = text.replace(/\s+/g, "");
  if (isRideGroupId(compact)) return { ref: compact };
  if (isTypedJoinCode(text)) return { ref: normalizeJoinCode(text) };
  return null;
}

export function groupInviteShareText(input: {
  title: string;
  url: string;
  code?: string;
  profileUrl?: string;
  visibility?: "public" | "private";
  when?: string;
  meetingPoint?: string;
}): string {
  const lines = [`Zusammen raus: ${input.title}`, input.url];
  if (input.visibility === "public") {
    const typed = normalizeJoinCode(input.code ?? "");
    if (typed.length === RIDE_GROUP_JOIN_CODE_LEN) {
      lines.push(`Code: ${typed}`);
    }
  }
  if (input.when?.trim()) {
    lines.push(input.when.trim());
  }
  if (input.meetingPoint?.trim()) {
    lines.push(`Treffpunkt: ${input.meetingPoint.trim()}`);
  }
  if (input.profileUrl?.trim()) {
    lines.push("", `Mein Profil: ${input.profileUrl.trim()}`);
  }
  const vis =
    input.visibility === "public"
      ? "Freigegeben: Link oder Code reicht. Die Gruppe steht auf dem Platz und als Treffen-Pin auf der Karte."
      : "Privat: nur wer diesen Link hat, kann beitreten. Nicht gelistet.";
  lines.push("", vis);
  return lines.join("\n");
}

export function publicProfileShareUrl(handle: string, origin?: string): string | undefined {
  const h = handle.trim().toLowerCase().replace(/[^a-z0-9_]/g, "");
  if (!h) return undefined;
  const base = (origin || siteOrigin() ||
    (typeof window !== "undefined" ? window.location.origin : "")
  ).replace(/\/$/, "");
  if (!base) return `/u/${h}`;
  return `${base}/u/${h}`;
}

export function inviteWindowOpen(
  payload: RideGroupInvitePayload,
  now = new Date()
): boolean {
  const start = new Date(payload.start);
  const end = new Date(payload.end);
  return now.getTime() <= end.getTime();
}

File: mobile/lib/data/community/ride_group_invite.dart
import 'dart:convert';

import '../../core/config.dart';
import '../../domain/community/ride_group.dart';
import '../../domain/community/ride_group_policy.dart';
import '../../domain/saved_route.dart';
import '../../domain/saved_route_note.dart';
import '../../domain/tours/tour_akte.dart';
import 'tour_share_codec.dart';

/// Teilbarer Einladungslink für „Zusammen raus“.
///
/// HTTPS (WhatsApp / Messages / Copy):
///   `https://aetherride.vercel.app/library?group=<id>&g=<token>`
/// Custom scheme (App öffnen):
///   `aetherride://platz?group=<id>&g=<token>`
///
/// `g` trägt Titel, Tour-Id und Fenster — kein Roster, keine Fake-Mitglieder.
/// Join-Code bleibt intern. Alte `?group=ABC234`-Links gelten weiter.
abstract final class RideGroupInvite {
  /// Share-Origin: Emulator/localhost wird zur Production, sonst bricht WhatsApp.
  static String shareOrigin({String? origin}) {
    final raw = (origin ?? AppConfig.apiBaseUrl).replaceAll(RegExp(r'/$'), '');
    if (raw.contains('localhost') ||
        raw.contains('127.0.0.1') ||
        raw.contains('10.0.2.2')) {
      return AppConfig.productionApiBaseUrl;
    }
    return raw;
  }

  static String httpsUrl({
    String? groupId,
    String? code,
    String? token,
    String? origin,
  }) {
    final base = shareOrigin(origin: origin);
    final ref = (groupId != null && groupId.trim().isNotEmpty)
        ? groupId.trim()
        : _code(code ?? '');
    final q = StringBuffer('$base/library?group=$ref');
    if (token != null && token.isNotEmpty) q.write('&g=$token');
    return q.toString();
  }

  static String? profileHttpsUrl({required String handle, String? origin}) {
    final h = handle.trim().toLowerCase().replaceAll(RegExp(r'[^a-z0-9_]'), '');
    if (h.isEmpty) return null;
    return '${shareOrigin(origin: origin)}/u/$h';
  }

  static String customSchemeUrl({
    String? groupId,
    String? code,
    String? token,
  }) {
    final ref = (groupId != null && groupId.trim().isNotEmpty)
        ? groupId.trim()
        : _code(code ?? '');
    final q = StringBuffer('aetherride://platz?group=$ref');
    if (token != null && token.isNotEmpty) q.write('&g=$token');
    return q.toString();
  }

  static String shareText({
    required String title,
    required String url,
    String? appUrl,
    String? code,
    String? profileUrl,
    RideGroupVisibility visibility = RideGroupVisibility.private,
    String? when,
    String? meetingPoint,
  }) {
    final buf = StringBuffer()
      ..writeln('Zusammen raus: $title')
      ..writeln(url);
    final app = appUrl?.trim() ?? '';
    if (app.isNotEmpty && app != url) {
      buf.writeln(app);
    }
    if (visibility == RideGroupVisibility.public) {
      final typed = code == null ? '' : RideGroupPolicy.normalizeJoinCode(code);
      if (typed.length == RideGroupPolicy.joinCodeLen) {
        buf.writeln('Code: $typed');
      }
    }
    if (when != null && when.trim().isNotEmpty) {
      buf.writeln(when.trim());
    }
    if (meetingPoint != null && meetingPoint.trim().isNotEmpty) {
      buf.writeln('Treffpunkt: ${meetingPoint.trim()}');
    }
    if (profileUrl != null && profileUrl.trim().isNotEmpty) {
      buf
        ..writeln()
        ..writeln('Mein Profil: ${profileUrl.trim()}');
    }
    buf
      ..writeln()
      ..write(
        visibility == RideGroupVisibility.public
            ? 'Freigegeben: Link oder Code reicht. '
                'Die Gruppe steht auf dem Platz und als Treffen-Pin auf der Karte.'
            : 'Privat: nur wer diesen Link hat, kann beitreten. '
                'Nicht gelistet.',
      );
    return buf.toString();
  }

  static String encode(
    RideGroup group, {
    SavedRouteEntry? route,
    SavedRouteMeta? meta,
    Map<String, dynamic>? tour,
  }) {
    Map<String, dynamic>? nextTour = tour;
    if (nextTour == null &&
        route != null &&
        needsMemberTrack(
          savedRouteId: group.savedRouteId,
          catalogTourId: group.catalogTourId ?? meta?.catalogTourId,
        )) {
      nextTour = buildTourSharePayload(
        route,
        meta: meta ?? SavedRouteMeta.empty,
      );
    }
    String token = _encodeBody(group, nextTour);
    if (token.length > 2400 &&
        nextTour != null &&
        nextTour['track'] != null) {
      nextTour = Map<String, dynamic>.from(nextTour)
        ..['includeTrack'] = false
        ..remove('track');
      token = _encodeBody(group, nextTour);
    }
    if (token.length > 2400) {
      token = _encodeBody(group, null);
    }
    return token;
  }

  static String _encodeBody(RideGroup group, Map<String, dynamic>? tour) {
    return _toBase64Url(
      jsonEncode({
        'v': 1,
        'kind': 'group',
        'id': group.id,
        'code': _code(group.joinCode),
        'title': group.title,
        'savedRouteId': group.savedRouteId,
        if (group.catalogTourId != null) 'catalogTourId': group.catalogTourId,
        'hostUserId': group.hostUserId,
        'start': group.startWindowStart.toUtc().toIso8601String(),
        'end': group.startWindowEnd.toUtc().toIso8601String(),
        if (tour != null) 'tour': tour,
      }),
    );
  }

  static RideGroupInvitePayload? decode(String token) {
    final raw = _fromBase64Url(token);
    if (raw == null) return null;
    try {
      final data = jsonDecode(raw);
      if (data is! Map) return null;
      if (data['v'] != 1 || data['kind'] != 'group') return null;
      final code = _code('${data['code'] ?? ''}');
      final title = '${data['title'] ?? ''}'.trim();
      final route = '${data['savedRouteId'] ?? ''}'.trim();
      final id = '${data['id'] ?? ''}'.trim();
      if (code.length != RideGroupPolicy.joinCodeLen ||
          title.isEmpty ||
          route.isEmpty ||
          id.isEmpty) {
        return null;
      }
      final start = DateTime.tryParse('${data['start']}');
      final end = DateTime.tryParse('${data['end']}');
      if (start == null || end == null) return null;
      return RideGroupInvitePayload(
        id: id,
        code: code,
        title: title,
        savedRouteId: route,
        catalogTourId: (data['catalogTourId'] as String?)?.trim(),
        hostUserId: '${data['hostUserId'] ?? ''}'.trim(),
        start: start.toUtc(),
        end: end.toUtc(),
        tour: parseTourShareMap(data['tour']),
      );
    } catch (_) {
      return null;
    }
  }

  static String _code(String raw) => raw.trim().toUpperCase();

  /// Paste from WhatsApp / Messages: HTTPS, App-Scheme, or bare id.
  /// Prefers the URL that carries `g=` (private groups need the token).
  static PlatzPastedJoin? parsePastedJoin(String raw) {
    final text = raw.trim();
    if (text.isEmpty) return null;
    PlatzPastedJoin? withoutToken;
    final urlRe = RegExp(
      r'(?:https?://|aetherride://)[^\s]+',
      caseSensitive: false,
    );
    for (final m in urlRe.allMatches(text)) {
      final uri = Uri.tryParse(m.group(0)!);
      if (uri == null) continue;
      final group =
          (uri.queryParameters['group'] ?? uri.queryParameters['code'] ?? '')
              .trim();
      if (group.isEmpty) continue;
      final token = uri.queryParameters['g']?.trim();
      final hit = PlatzPastedJoin(
        code: group,
        token: (token == null || token.isEmpty) ? null : token,
      );
      if (hit.token != null) return hit;
      withoutToken ??= hit;
    }
    if (withoutToken != null) return withoutToken;
    final compact = text.replaceAll(RegExp(r'\s+'), '');
    if (RideGroupPolicy.isGroupId(compact)) {
      return PlatzPastedJoin(code: compact);
    }
    if (RideGroupPolicy.isTypedJoinCode(text)) {
      return PlatzPastedJoin(code: RideGroupPolicy.normalizeJoinCode(text));
    }
    return null;
  }
}

class PlatzPastedJoin {
  const PlatzPastedJoin({required this.code, this.token});

  final String code;
  final String? token;
}

class RideGroupInvitePayload {
  const RideGroupInvitePayload({
    required this.id,
    required this.code,
    required this.title,
    required this.savedRouteId,
    required this.start,
    required this.end,
    this.catalogTourId,
    this.hostUserId = '',
    this.tour,
  });

  final String id;
  final String code;
  final String title;
  final String savedRouteId;
  final String? catalogTourId;
  final String hostUserId;
  final DateTime start;
  final DateTime end;
  final Map<String, dynamic>? tour;

  bool windowOpen(DateTime now) => !now.isAfter(end);
}

String _toBase64Url(String json) {
  return base64Url.encode(utf8.encode(json)).replaceAll('=', '');
}

String? _fromBase64Url(String token) {
  try {
    var b64 = token.replaceAll('-', '+').replaceAll('_', '/');
    final pad = b64.length % 4 == 0 ? 0 : 4 - (b64.length % 4);
    b64 = b64.padRight(b64.length + pad, '=');
    return utf8.decode(base64.decode(b64));
  } catch (_) {
    return null;
  }
}
